README.markdown

```diff
@@ -6669,6 +6671,46 @@ This feature was first introduced in the `v0.7.3` release.
 
 [Back to TOC](#nginx-api-for-lua)
 
+ngx.shared.DICT.stats
+---------------------
+**syntax:** *free_page_bytes = ngx.shared.DICT:stats()*
```

Comment:
Hmm, I thought you would return the total size here as well. Otherwise the method name `stats` is just too general for this.

Reply:
I think the total size is fun to watch but it's not useful for alerting the memory exhaustion of a shared dictionary.
So I'd like to avoid wasting to calculate unused metric here.

If the total size really needed, I'd like a separate method or make it optional like
`ngx.shared.DICT:stats(want_total_size)`

```
free_page_bytes = ngx.shared.DICT:stats()
free_page_bytes, total_size = ngx.shared.DICT:stats(true)
```

How about that?

Reply:
@hnakamur The boolean argument is not readable at all. How about 2 separate methods, `free_space()` and `capacity()`?

Reply:
I agree on 2 separate methods and the name `free_space`, but I think `capacity` is misleading because it reminds the limit of the size of the shared dict, not the current used bytes.
How about `usage()` or `used_bytes()` ?

Reply:
I've read code again to know how to implement the function for the total size and I realized I can only get an approximate value.

For a key value pair which belongs to a slot for size less than or equal to 2048 bytes, the total size used can be retrieved from `total` field of `ngx_slab_stat_t` struct. https://github.com/nginx/nginx/blob/release-1.13.5/src/core/ngx_slab.h#L26

However, for key value pair with management data whose size is greater than 2048 bytes, pages are assigned at https://github.com/nginx/nginx/blob/release-1.13.5/src/core/ngx_slab.c#L191-L206
and the request size is not recorded in `total` field of `ngx_slab_stat_t` struct. 
Instead, the key length and value length are recorded in `key_len` and `value_len` fields of `ngx_http_lua_shdict_node_t` struct.
https://github.com/openresty/lua-nginx-module/blob/97fbeb0bef1aa85e758210d58063376de8eaed31/src/ngx_http_lua_shdict.c#L1211-L1227

To read the `key_len` and `value_len` fields of `ngx_http_lua_shdict_node_t`, I think we need to iterate through all key value pairs in the shared dict, so it becomes a costly operation.
We would need the same caution as in `get_keys` method.
https://github.com/openresty/lua-nginx-module#ngxshareddictget_keys

But I'll avoid to use the `usage` method if it may lock the shared dictionary for significant amount of time and block Nginx worker processes trying to access the dictionary.

What do you think?

Reply:
@hnakamur capacity indeed means the total size specified in the `lua_shared_dict` config directive.

Reply:
@agentzh Oh I see. I've added `capacity` method and updated tests and docs.
Could you review it?
Thanks!